src/der.rs

```diff
@@ -104,9 +117,116 @@ pub(crate) fn expect_tag<'a>(
+            let combined = (length_byte_one << 16) | (length_byte_two << 8) | length_byte_three;
+            if combined < LONG_FORM_LEN_TWO_BYTES_MAX {
+                return Err(Error::BadDer); // Not the canonical encoding.
+            }
+            combined
+        }
+        LONG_FORM_LEN_FOUR_BYTES => {
+            let length_byte_one = usize::from(input.read_byte()?);
+            let length_byte_two = usize::from(input.read_byte()?);
+            let length_byte_three = usize::from(input.read_byte()?);
+            let length_byte_four = usize::from(input.read_byte()?);
+            let combined = (length_byte_one << 24)
+                | (length_byte_two << 16)
+                | (length_byte_three << 8)
+                | length_byte_four;
+            if combined < LONG_FORM_LEN_THREE_BYTES_MAX {
+                return Err(Error::BadDer); // Not the canonical encoding.
+            }
+            combined
+        }
+        _ => {
+            return Err(Error::BadDer); // We don't support longer lengths.
+        }
+    };
+
+    if length >= size_limit {
+        return Err(Error::BadDer); // The length is larger than the caller accepts.
+    }
+
+    let inner = input.read_bytes(length)?;
+    Ok((tag, inner))
 }
 
+// Long-form DER encoded lengths of two bytes can express lengths up to the following limit.
+//
+// The upstream ring::io::der::read_tag_and_get_value() function limits itself to up to two byte
+// long-form DER lengths, and so this limit represents the maximum length that was possible to
+// read before the introduction of the read_tag_and_get_value_limited function.
+pub(crate) const TWO_BYTE_DER_SIZE: usize = LONG_FORM_LEN_TWO_BYTES_MAX;
+
+// DER Tag identifiers have two forms:
+// * Low tag number form (for tags values in the range [0..30]
+// * High tag number form (for tag values in the range [31..]
+// We only support low tag number form.
+const HIGH_TAG_RANGE_START: u8 = 31;
+
+// DER length octets have two forms:
+// * Short form: 1 octet supporting lengths between 0 and 127.
+// * Long definite form: 2 to 127 octets, number of octets encoded into first octet.
+const SHORT_FORM_LEN_MAX: u8 = 128;
+
+// Leading octet for long definite form DER length expressed in second byte.
+const LONG_FORM_LEN_ONE_BYTE: u8 = 129;
+
+// Maximum size that can be expressed in a one byte long form len.
+const LONG_FORM_LEN_ONE_BYTE_MAX: usize = 1 << 8;
+
+// Leading octet for long definite form DER length expressed in subsequent two bytes.
+const LONG_FORM_LEN_TWO_BYTES: u8 = 130;
+
+// Maximum size that can be expressed in a two byte long form len.
+const LONG_FORM_LEN_TWO_BYTES_MAX: usize = 1 << (8 * 2);
+
+// Leading octet for long definite form DER length expressed in subsequent three bytes.
+const LONG_FORM_LEN_THREE_BYTES: u8 = 131;
+
+// Maximum size that can be expressed in a three byte long form len.
+const LONG_FORM_LEN_THREE_BYTES_MAX: usize = 1 << (8 * 3);
```

Comment:
This is actually one more than the maximum. ie, three bytes can encode `0xff_ff_ff` but the value here is `0x1_00_00_00`. The code is still correct because the bound is half-open (`combined < LONG_FORM_LEN_THREE_BYTES_MAX `)

Reply:
Good catch :-)

I've fixed the constants and updated the comparison to use `<=`. I think expressing it this way will make it easier to reason about.